scipy/cluster/hierarchy.py

```diff
@@ -1052,29 +1027,33 @@ def linkage(y, method='single', metric='euclidean', optimal_ordering=False):
             raise ValueError("Method '{}' requires the distance metric "
                              "to be Euclidean".format(method))
         if y.shape[0] == y.shape[1] and np.allclose(np.diag(y), 0):
-            if np.all(y >= 0) and np.allclose(y, y.T):
+            if xp.all(y >= 0) and xp.allclose(y, y.T):
```

Comment:
Just to be clear, array_api_compat does, by design, include all functions from the wrapped module, even those not in the array API. So when `xp` is `numpy`, `xp.allclose` will work here. If you want to ensure portability across any potential array API namespace you need to test with `numpy.array_api`. 

It looks like allclose is in numpy, cupy, and torch. See also https://github.com/data-apis/array-api/issues/170

Reply:
yeah so that's one big question I suppose. If we are saying we are only supporting the strict Array API, hence yes test with `numpy.array_api` (which I forgot to do in my last pass), or we say we support a set of libraries (which happen to be the ones in `array_api_compat`, i.e. pushing some responsibility there).

Reply:
> If we are saying we are only supporting the strict Array API, hence yes test with `numpy.array_api`

This. We should test with `numpy.array_api`. That was always the plan, and there will be needs - e.g., @tylerjereddy already indicated that there is interest to test a LANL-internal accelerator library. It's also the only thing that is well-defined; sticking to whatever happens to be the intersection of the NumPy, PyTorch and CuPy APIs that seems to pass our test suite is a bit of a moving target.